remove need for torch in torch.package

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74929
* __->__ #72961
* #72237

Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

PaliC committed:
remove need for torch in torch.package

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"




[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


There is still a bug with zipfile/directory naming but worth taking a look to see if it's the right approach.

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


There is still a bug with zipfile/directory naming but worth taking a look to see if it's the right approach.

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

d4l3k commented on torch/package/package_importer.py:
```diff
@@ -178,49 +271,17 @@ def load_pickle(self, package: str, resource: str, map_location=None) -> Any:
-                )
-            elif typename == "reduce_package":
+            if self.torch_is_available:
+                module = self.persistent_id_torch(typename, data)
```
same thing here, we could have two loader methods with identical interfaces that are selected at import time, that way we can cut down on these if else checks

d4l3k commented on torch/package/package_importer.py:
```diff
@@ -27,8 +23,29 @@
+    import torch
+    from ._directory_reader_torchscript import TorchScriptDirectoryReader
+    from ._zip_file_torchscript import TorchScriptPackageZipFileReader
+    from torch.serialization import _get_restore_location
```
these conditional global imports get quite fragile, if we can wrap them with fallbacks at import time that seems much better

d4l3k reviewed (commented):
I left some general comments in a couple spots but they apply to both importer/exporter. Would be nice to clean this interface up since the if/else statements are fairly fragile.

In terms of testing it would be nice to run the same test suite on both w/ torch and without. We had to do this for TorchX here: https://github.com/pytorch/torchx/blob/main/torchx/schedulers/test/kubernetes_scheduler_test.py#L554-L569

You can mock torch to not import and then reload `torch.package` to run the tests without `torch` present to ensure everything works E2E

PaliC commented on torch/package/package_exporter.py:
```diff
@@ -184,7 +189,7 @@ def __init__(
         f: Union[str, Path, BinaryIO],
         importer: Union[Importer, Sequence[Importer]] = sys_importer,
-        zip_file_reader_type: Type[PackageZipFileWriter] = TorchScriptPackageZipFileWriter
+        use_torch: bool = False,
```
Yeah the idea here is to have users specify this if needed as the 
use torch exporter => use torchless importer 
use case exists, but there isn't a good way to do it. We could just say it doesn't matter as it does make things cleaner.

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
add unit tests for the new zipfile classes on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx.

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

d4l3k reviewed (approved):
Generally this looks good (mostly just moving stuff around) there's a bit of a weird pattern of having a second method that can be overridden. If we can move the logic there into that method I think it would clean this up

instead of
```py
def foo(self, ...): 
  return None

def _foo(self, ...):
  ...
  obj = self.foo()
  if obj is not None: 
    return obj
  # more logic
```
you can do
```py
def foo(self, ...): 
  # more logic
  return ...

def _foo(self, ...):
  ...
  return self.foo()
  
```


Side note: `git diff --color-moved` is really a godsend for reviewing this code

PaliC commented on torch/package/package_exporter.py:
```diff
@@ -902,253 +59,4 @@ def _persistent_id(self, obj):
-            if _gate_torchscript_serialization and isinstance(
-                obj, torch.jit.RecursiveScriptModule
-            ):
-                raise Exception(
```
Yep, given it works internally for some teams, we have concluded that the flag is no longer needed.

PaliC commented on torch/package/package_exporter_no_torch.py:
```diff
@@ -0,0 +1,1111 @@
+            _ModuleProviderAction.DENY, allow_empty=True
+        )
+
+    def persistent_id(self, obj):
```
Yeah! Though I can refactor it to just move it into save_pickle.

PaliC replied:
Scratch that .... I forgot that torch.package (non torch bits) is basically a wrapper around pickle. We can't just dump it into save_pickle. I clean it up a bit with the pattern you mentioned

PaliC replied:
Scratch that .... I forgot that torch.package (non torch bits) is basically a wrapper around pickle. We can't just dump it into save_pickle. I clean it up a bit with the pattern you mentioned

PaliC replied:
Sorry for the email spam. Since we want __reduce_pacakge__ support, and user persistent_id functionality should override ours. I think this is fine. This is also more or less the pattern pickle.py has

persistent_id - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L540

__reduce__ - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L580

PaliC replied:
Sorry for the email spam. Since we want __reduce_pacakge__ support, and user persistent_id functionality should override ours. I think this is fine. This is also more or less the pattern pickle.py has

persistent_id - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L540

__reduce__ - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L580

PaliC commented on torch/package/package_importer.py:
```diff
@@ -1,160 +1,76 @@
+                    self.restore_location,
+                )
+            storage = self.loaded_storages[key]
+            # TODO: Once we decide to break serialization FC, we can
```
I don't think so ... I'm not exactly sure what Zach was referring to

PaliC commented on torch/package/package_importer.py:
```diff
@@ -1,160 +1,76 @@
+            self.restore_location = None
+            self.loaded_storages = None
+
+    # TODO: load_pickle to reduce the repeated code between this and the non-torch version
```
That's true, I'll turn it into a bootcamp task.

PaliC commented on torch/package/package_importer_no_torch.py:
```diff
@@ -0,0 +1,665 @@
+    """
+    modules: Dict[str, types.ModuleType]
+
+    def get_zip_reader(self, file_or_buffer):
```
Oh thanks for the catch! I missed this

PaliC commented on torch/package/package_importer_no_torch.py:
```diff
@@ -0,0 +1,665 @@
+            module = self.persistent_load(typename, data)
+            if module is not None:
+                return module
+            if typename == "reduce_package":
```
Yeah let's do that!

PaliC replied:
Sorry I misread, we always want the reduce_package behavior, so no.

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).

PaliC committed:
Update on "remove need for torch in torch.package"


Remove the requirement of torch in PackageExporter and PackageImporter. Now both of these modules work without torch, however, they can use torch for the necessary custom operations like in the case of torchscript and torch.fx. We do this by extending two endpoints similar to pickle call persistent_id and __reduce_pacakge__

This diff also adds unit tests for the torchless/and normal version of PackageExporter/PackageImporter

Differential Revision: [D34918137](https://our.internmc.facebook.com/intern/diff/D34918137)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34918137).
